Preparations for new Tracing/Replay implementation

daumayr opened the pull request:
This PR contains preparation for the new tracing & replay.
The custom ByteBuffer is only temporary and will be removed in the next PR.
Please note that replay is not functional as of this PR.

Due to the split into two tracing variants, a lot of code moved to different files.

smarr committed:
Optimize tracing

- we know it is a TracingActor at this point, no need for a virtual call to `getId()`
- fix seemingly incorrect sync on fullBuffers
- prefer isEmpty over size()
- fix check for whether there is something in the buffer
- position() should be the correct thing, untested though
- use unsafe to avoid separate byte writes
- avoid synchronizing on a synchronized data structure
- add additional methods to Unsafe dummy.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr commented on src/som/primitives/SystemPrims.java:
```diff
@@ -45,13 +45,17 @@
 import tools.SourceCoordinate;
+import tools.concurrency.ActorExecutionTrace;
+import tools.concurrency.TraceParser;
+import tools.concurrency.TracingBackend;
```
@daumayr Please fix these issues.

smarr commented:
After fixing the infrastructure issues (which seemed trivial and I bet it would have been easy to figure out), the Kompos tests are failing.

How do you want to play this? Are you going to look into the issues?

daumayr committed:
Cleanup in Files

make stream creation simpler

daumayr committed:
Remove -atcfg flag and add -sid flag for short ids

The old actor config flags are not used any longer.

daumayr committed:
Add notion of external messages for replay

External messages are going to be use to represent information coming in from outside of the system.

The capture the external non-determinism.
And adds data wrappers for some of the payload.

This is just an initial change, they are not yet used.

daumayr committed:
Rename MEDEOR_TRACING to KOMPOS_TRACING

Use the name of the protocol that is supported by this, not the subsystem that implements it.

daumayr committed:
Split Tracing into Kompos and replay support

daumayr committed:
Update tracing and replay

daumayr committed:
Trace system time and add datasource interface

smarr commented on src/som/primitives/StringPrims.java:
```diff
@@ -150,6 +150,7 @@ protected boolean isTaggedWithIgnoringEagerness(final Class<?> tag) {
       }
     }
 
+    @TruffleBoundary
```
Hmm, why did you add a boundary here?
The code looks fine. It should compile, no?

daumayr replied:
It's probably a relict of changes that are not in this PR.
If I have an issue with this at some point , I'll solve it differently.

smarr replied:
I just wrote a microbenchmark to see whether and how it compiles.
And I looked at the IGV graphs. It compiles surprisingly good, despite the three times bounds checking (SOMns, Java string, char array).

So, I would expect it to be safe to remove the boundary here. (will do)

smarr committed:
Move ActorExecutionTrace to replay package

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr commented on src/som/interpreter/actors/Actor.java:
```diff
@@ -298,16 +285,16 @@ private void execute(final EventualMessage msg,
-              msg.getTargetSourceSection());
-        }
-        msg.execute();
-      } finally {
```
Is the removal of this finally intentional? Seems to be the safer to use a finally here.

smarr committed:
Fix expected test results

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr commented:
@daumayr ok, it is merged. Looking forward to the other parts.
